# <font face = 'Impact' color = '#FFAEBC' > Getting Ready with our Second Mini-Project <font/>
#### <font face = 'Times New Roman' color = '#A16AE8'> Author and Trainer: Paolo Hilado MSc. (Data Science)<font/>
In this notebook, we will be importing packages and use them to load and explore data sets. We will be using the pandas package to load datasets and qdesc to explore both categorical and numerical variables. Functions to be used include pd.read_excel, pd.read_csv, head, qd.desc, qd.normcheck_dashboard, qd.freqdist_a, and qd.freqdist_to_excel.

# <font face = 'Impact' color = '#FFAEBC' > Exploring Sleep Hours during Weeknights <font/>

In [ ]:
pip install qdesc

In [ ]:
# Importing the necessary packages for us to use certain functions.
import pandas as pd
import numpy as np
import qdesc as qd

In [ ]:
# Loading a data set (SleepData.xlsx) and storing it into a variable.
df = pd.read_excel('SleepData.xlsx')

In [ ]:
# Check the first 5 rows of the dataframe
df.head()

,Sex,Age,MaritalStatus,Employment,sleep(hrs)/weeknights
0,Male,30,Single,Permanent,8.85
1,Female,58,Married,Permanent,8.76
2,Female,60,Single,Permanent,8.49
3,Male,43,Single,Permanent,8.85
4,Female,43,Single,Permanent,8.73


In [ ]:
# Check the information of the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 671 entries, 0 to 670
Data columns (total 5 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Sex                    671 non-null    object 
 1   Age                    671 non-null    int64  
 2   MaritalStatus          671 non-null    object 
 3   Employment             671 non-null    object 
 4   sleep(hrs)/weeknights  671 non-null    float64
dtypes: float64(1), int64(1), object(3)
memory usage: 26.3+ KB


In [ ]:
# Using the desc function from the qdesc package to perform descriptive analysis.
qd.desc(df)

,count,mean,std,median,MAD,min,max,AD_stat,5% crit_value
Age,671.0,49.80,19.36,49.00,17.00,18.00,84.00,7.38,0.78
sleep(hrs)/weeknights,671.0,7.63,1.39,8.34,0.98,4.52,9.88,35.70,0.78


In [ ]:
# Generate descriptive analysis for sleep hours when grouped by sex.
df.groupby('Sex')['sleep(hrs)/weeknights'].describe()

,count,mean,std,min,25%,50%,75%,max
Sex,,,,,,,,
Female,350.0,7.672943,1.358217,4.52,6.3475,8.33,8.88,9.88
Male,321.0,7.574860,1.425584,4.66,6.2100,8.36,8.89,9.64


The analysis shows that both males and females sleep an average of about 7.6–7.7 hours on weeknights, with very small differences between the groups. The median sleep is ~8.3 hours for both, indicating most respondents get adequate rest. Variability is similar, though males show slightly higher spread. Overall, sleep patterns are consistent across sexes.



In [ ]:
# Generate descriptive analysis for sleep hours when grouped by employment.
df.groupby('Employment')['sleep(hrs)/weeknights'].describe()

,count,mean,std,min,25%,50%,75%,max
Employment,,,,,,,,
Permanent,352.0,8.868438,0.322398,7.96,8.64,8.86,9.0925,9.88
Probationary,319.0,6.255078,0.604951,4.52,5.86,6.22,6.6400,7.90


The results show a clear difference in sleep patterns by employment status. Permanent employees sleep an average of 8.87 hours, with little variation, while probationary employees average only 6.26 hours, with greater variability. This suggests that permanent employees generally get sufficient rest, whereas probationary employees tend to have shorter and less consistent sleep on weeknights.

In [ ]:
# Generate descriptive analysis for sleep hours when grouped by marital status.
df.groupby('MaritalStatus')['sleep(hrs)/weeknights'].describe()

,count,mean,std,min,25%,50%,75%,max
MaritalStatus,,,,,,,,
Married,333.0,7.667838,1.396639,4.66,6.2500,8.44,8.93,9.64
Single,338.0,7.584822,1.385605,4.52,6.2725,8.26,8.85,9.88


The results show that married respondents sleep an average of 7.67 hours, while single respondents average 7.58 hours—a very small difference. Both groups have similar variability and medians (~8.3–8.4 hours), suggesting that marital status does not significantly affect weeknight sleep patterns.

In [ ]:
# Use the desc function from the qdesc package to explore the data for assumption of normality.
desc_df = qd.desc(df)
desc_df

,count,mean,std,median,MAD,min,max,AD_stat,5% crit_value
Age,671.0,49.80,19.36,49.00,17.00,18.00,84.00,7.38,0.78
sleep(hrs)/weeknights,671.0,7.63,1.39,8.34,0.98,4.52,9.88,35.70,0.78


In [ ]:
# Slicing the data frame to only include the single.
single = df[df['MaritalStatus'] == 'Single']
single

,Sex,Age,MaritalStatus,Employment,sleep(hrs)/weeknights
0,Male,30,Single,Permanent,8.85
2,Female,60,Single,Permanent,8.49
3,Male,43,Single,Permanent,8.85
4,Female,43,Single,Permanent,8.73
6,Male,45,Single,Permanent,8.36
...,...,...,...,...,...
663,Female,61,Single,Probationary,6.51
664,Male,39,Single,Probationary,5.74
667,Female,76,Single,Probationary,4.90
669,Female,25,Single,Probationary,5.18


In [ ]:
# Confirm that the subset would only include those who are single.
single ['MaritalStatus'].value_counts()

,count
MaritalStatus,
Single,338


In [ ]:
# Use the qdesc package to explore the data (only those who are Single) for assumption of normality.
qd.desc(single)

,count,mean,std,median,MAD,min,max,AD_stat,5% crit_value
Age,338.0,50.74,19.11,50.00,16.00,18.00,84.00,3.05,0.78
sleep(hrs)/weeknights,338.0,7.58,1.39,8.26,1.01,4.52,9.88,16.56,0.78


In [ ]:
# Slicing the data frame to only include the married.
married = df[df['MaritalStatus'] == 'Married']
married

,Sex,Age,MaritalStatus,Employment,sleep(hrs)/weeknights
1,Female,58,Married,Permanent,8.76
5,Female,32,Married,Permanent,8.70
10,Female,29,Married,Permanent,9.05
11,Female,31,Married,Permanent,9.41
12,Male,38,Married,Permanent,8.94
...,...,...,...,...,...
661,Female,67,Married,Probationary,7.20
662,Male,20,Married,Probationary,6.38
665,Male,39,Married,Probationary,6.22
666,Female,59,Married,Probationary,6.15


In [ ]:
# Confirm that the subset is all married.
married ['MaritalStatus'].value_counts()


,count
MaritalStatus,
Married,333


In [ ]:
# Use the qdesc package to explore the data for assumption of normality.
qd.desc(married)

,count,mean,std,median,MAD,min,max,AD_stat,5% crit_value
Age,333.0,48.86,19.6,48.00,17.00,18.00,84.00,4.71,0.78
sleep(hrs)/weeknights,333.0,7.67,1.4,8.44,0.88,4.66,9.64,19.32,0.78


In [ ]:
# Generate a frequency distribution for a specified variable: Sex
df["Sex"].value_counts()

,count
Sex,
Female,350
Male,321


The frequency distribution shows that the dataset has 350 females and 321 males. This indicates a fairly balanced sample, with females making up a slightly larger portion of the respondents.

In [ ]:
# Generate a frequency distribution for a specified variable: Employment
df["Employment"].value_counts()

,count
Employment,
Permanent,352
Probationary,319


The frequency distribution shows that there are 352 permanent employees and 319 probationary employees in the dataset. This indicates a nearly balanced distribution, with a slightly higher number of permanent employees.

In [ ]:
# Generate a frequency distribution for a specified variable: Marital Status
df["MaritalStatus"].value_counts()


,count
MaritalStatus,
Single,338
Married,333


The frequency distribution for marital status shows that the dataset is almost evenly divided, with 338 respondents being single and 333 respondents being married. This indicates a well-balanced representation between the two groups.

In [ ]:
# Generate frequency distribution for all the categorical features from the said data frame.
categorical = df.select_dtypes(include=["object", "category"]).columns
for col in categorical:
   counts = df[col].value_counts()
   percentages = df[col].value_counts(normalize=True) * 100
   print(pd.DataFrame({"Counts": counts, "Percentages": percentages}))
   print("\n")




        Counts  Percentages
Sex                        
Female     350    52.160954
Male       321    47.839046


               Counts  Percentages
MaritalStatus                     
Single            338    50.372578
Married           333    49.627422


              Counts  Percentages
Employment                       
Permanent        352    52.459016
Probationary     319    47.540984




The dataset shows a fairly balanced distribution across all categorical variables. In terms of sex, females make up 52.2% while males account for 47.8%, showing only a slight difference. Marital status is almost evenly split, with 50.4% single and 49.6% married respondents. Employment status is also balanced, with 52.5% being permanent employees and 47.5% probationary. Overall, the dataset provides a well-distributed representation across these categories, without any major skew toward a particular group.

In [ ]:
# Generating a frequency distribution but this time have it stored in an excel file and organized
# on a per sheet basis per categorical feature.
categorical= df.select_dtypes(include='object').columns
excel_file_path = 'frequency_distributions.xlsx'

with pd.ExcelWriter(excel_file_path) as writer:
    for col in categorical:
        counts = df[col].value_counts()
        percentages = df[col].value_counts(normalize=True) * 100
        freq_df = pd.DataFrame({'Counts': counts, 'Percentages': percentages})
        freq_df.to_excel(writer, sheet_name=col)

print(f"Frequency distributions saved to {excel_file_path}")


Frequency distributions saved to frequency_distributions.xlsx
